### PR TITLE
Add ranking metrics and Encoder Model for training

### PR DESCRIPTION
**Pull request by Naqu6**
Work Done:
- Added `rankings.py`, which contains codes for computing rankings
- Added `encoder.py` which wraps a Text Encoder and a Video Encoder for training
- Added tests for `rankings.py`
- Updated `loss.py`, separating similarity matrix building and the ranking loss function
- Updated `loss_tests.py` to account for change to loss function
- Added tests for `encoder.py`
- Updated tests for the Video and Text Encoder

### CODE REVIEW

File: videoretrieval/tests/rankings_tests.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_metric_computations(self):
+        """Tests computing metrics from ranked queries."""
+        mock_ranks = tf.constant([1, 1, 2, 5, 5, 7, 9, 8, 10, 11, 13, 15])
+        k_expected_recall_pairs = [
+            (1, 1/6),
+            (5, 5/12),
+            (10, 9/12),
+            (11, 10/12),
+            (15, 1.0)]
```
Comment (ozdemirbahadir) [inline videoretrieval/tests/rankings_tests.py:49]: k_expected_recall_pairs = [(1, 0), (5, 0), (10, 0), (11, 0), (15, 1.0)]

Is this intended?

File: videoretrieval/metrics/loss.py [hunk 2]
```diff
@@ -73,48 +73,36 @@ def build_similarity_matrix(
...
-        similarities_transpose - matching_similarities)
+        embedding_margin + similarities_transpose - matching_similarities)
 
     computed_similarities = computed_similarities *  (1 - tf.eye(batch_size))
 
     loss = tf.reduce_sum(
         computed_similarities) / (2 * (batch_size**2 - batch_size))
 
-    return loss
+    return loss
```
Comment (ozdemirbahadir) [inline videoretrieval/metrics/loss.py:108]: Add a new line.

File: videoretrieval/models/encoder.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            similarity_matrix = build_similarity_matrix(
+                video_results, text_results, mixture_weights, missing_experts)
+            loss = self.loss_fn(similarity_matrix, self.loss_hyperparameter_m)
+
+        gradients = gradient_tape.gradient(loss, self.trainable_variables)
+
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+
+        # It's wasteful to calculate ranking metrics for the entire train
+        # dataset, so we just mark the values as nan for keras.
```
Comment (ozdemirbahadir) [inline videoretrieval/models/encoder.py:97]: s/nan/NaN

File: videoretrieval/tests/loss_tests.py [hunk 4]
```diff
@@ -50,19 +50,21 @@ def test_loss_calculation_with_different_embeddings(self):
         ])]
 
         loss = bidirectional_max_margin_ranking_loss(
-            mock_perfect_video_embeddings,
-            mock_perfect_text_embeddings,
-            mock_mixture_weights,
-            mock_missing_experts,
+            build_similarity_matrix(
```
Comment (ozdemirbahadir) [inline videoretrieval/tests/loss_tests.py:53]: Assign this matrix to a variable as follows:

similarity_matrix = build_similarity_matrix(
  mock_perfect_video_embeddings,
  mock_perfect_text_embeddings,
  mock_mixture_weights,
  mock_missing_experts);

loss = bidirectional_max_margin_ranking_loss(similarity_matrix, 1.0)


ditto for other usages.

File: videoretrieval/tests/rankings_tests.py (hunk 1 above)
Comment (ozdemirbahadir) [reply videoretrieval/tests/rankings_tests.py:49]: Add ".0" for readability.

Comment (Naqu6) [reply videoretrieval/tests/rankings_tests.py:49]: Updated!

File: videoretrieval/models/encoder.py (hunk 3 above)
Comment (Naqu6) [reply videoretrieval/models/encoder.py:97]: Good catch! Thanks Bahadir
